CIF-1134 - Unified configuration of site settings

dplaton opened the pull request:
## Description

This is quite a big PR due to the fact that it contains both the UI implementation and the updates to the connector code.

Changes in the connector code:
- the `GraphqlResourceProviderFactory` now reads the configuration using the Sling Context-Aware Configuration API (falls back to reading the configuration from the root resource if the `cq:conf` property is not found)
- the `CatalogDataResourceProviderManagerImpl` now reads the configuration using the Sling Context-Aware Configuration API. It also listens to specific property changes in `/conf` to update the data roots 

The configuration UI:
- a whole new console available at `http://localhost:4502/libs/commerce/gui/content/configuration.html/conf` through which the user has to create the configuration
- The product binding wizard has been reduced to just three fields: Name, Title and the path to the context configuration.

## Related Issue
CIF-1201 - Implement the configuration wizard
CIF-1204 - Update the connector to use context-aware configurations

## Motivation and Context

Unify configuration of the CIF Connector and CIF Core Components

## How Has This Been Tested?

Intensive functional testing.
Unit tests for the Sling Models are WIP

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [X] All unit tests pass on CircleCi.
- [X] I ran all tests locally and they pass.

dplaton committed:
CIF-1134 - Unified configuration of site settings

Update the Bind Products Wizard to include the `cq:conf` field.
Update the Resource Provider Factory to use the configurations from `/conf`

dplaton committed:
CIF-1134 - Unified configuration of site settings

Update the Catalog Data Resource Provider Manager and dependant components to use data from `/conf`

dplaton committed:
CIF-1204 - Update the connector to use context-aware configurations

Fix a bug that was causing the resource provider to throw a NPE during creation

dplaton committed:
CIF-1134 - Unified configuration of site settings

Implement the configurations console for CIF

dplaton committed:
øMerge branch 'issue/CIF-1134' of github.com:adobe/commerce-cif-connector into issue/CIF-1134

dplaton committed:
CIF-1204 - Update the connector to use context-aware configurations

Implement the console for the CIF Cloud configuration

dplaton committed:
CIF-1204 - Update the connector to use context-aware configurations

Update the Graphql Connector to use context aware configurations and fallback to the resource tree properties.
Update unit tests.

dplaton committed:
@trivial Code formatting

dplaton committed:
CIF-1201 - Implement the configuration wizard

Update the CIF Configurations console so that it shows the commerce configurations

dplaton committed:
CIF-1201 - Implement the configuration wizard

Implement the configuration dialog (WIP)

dplaton committed:
CIF-1204 - Update the connector to use context-aware configurations

Fix unit tests that were still using the ConfigurationResourceResolver

dplaton committed:
Merge branch 'issue/CIF-1134' of github.com:adobe/commerce-cif-connector into issue/CIF-1134

dplaton committed:
CIF-1201 - Implement the configuration wizard

Implement the configuration dialog for the configuration page

dplaton committed:
CIF-1201 - Implement the configuration wizard

Implement the "Create" functionality (WIP)

dplaton committed:
CIF-1201 - Implement the configuration wizard

 Implement the "Create configuration" action.

dplaton committed:
CIF-1201 - Implement the configuration wizard

 Implement the "Delete" action (WIP)

dplaton committed:
CIF-1201 - Implement the configuration wizard

 Implement the "Create configuration folder" action (WIP)

dplaton committed:
CIF-1201 - Implement the configuration wizard

Implement the "Create configuration folder" feature (WIP)

dplaton committed:
Move the configuration to /conf/.../settings/cloudconfigs/commerce.

dplaton committed:
Update the configuration name to "cloudconfigs/commerce"

dplaton committed:
Merge branch 'master' into issue/CIF-1134

dplaton committed:
@trivial Fix formatting

dplaton committed:
Move the configurations to `/conf/../settings/cloudconfig/commerce` and get rid of the servlet that creates the configuration folder.

dplaton committed:
Import the `javax.annotation` package so that we can use Sling Models

dplaton committed:
Remove the ConfigurationResourceResolver from the unit tests.

dplaton committed:
Update the console so you can create sub-configurations. The inheritance is on by default for now.

dplaton committed:
@trivial Delete unused client library

dplaton committed:
Add the "preview column" feature in the configuration console.

dplaton committed:
Add the "preview column" feature in the configuration console.

dplaton committed:
Remove unused properties from the Bind Product Tree wizard

dplaton committed:
Remove unused properties from the Bind Product Tree wizard

dplaton committed:
Update the CatalogDataResourceProviderManagerImpl to also listen to changes under `/conf` and update the data roots accordingly.

dplaton committed:
Fix the unit test failing
Update the subservice name

dplaton committed:
@trivial Remove unused reference to `ConfigurationResourceResolver`

dplaton committed:
Merge branch 'master' into issue/CIF-1134

dplaton committed:
Fix the sub-service name in the configuration file.

dplaton committed:
Merge branch 'master' into issue/CIF-1134

dplaton committed:
@trivial Remove the Configurations Datasource since we don't use it anymore

cjelger commented on bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/gui/Constants.java:
```diff
@@ -0,0 +1,38 @@
+
+    public static String COMMERCE_BUCKET_PATH = CONF_CONTAINER_BUCKET_NAME + "/" + COMMERCE_BUCKET_NAME;
+
+    private Constants() {
```
This is probably not needed, right?

dplaton replied:
Ah, yes, I forgot to remove it, thanks.

cjelger commented on bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/gui/components/configuration/ConfigurationColumnPreview.java:
```diff
@@ -0,0 +1,91 @@
+/*
+ * ******************************************************************************
+ *  *
```
The header is not correct, it has "twice" the left column with `*` characters.

cjelger commented on bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/gui/components/configuration/ConfigurationColumnPreview.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    public String getTitle() {
+        return properties.get("jcr:title", itemResource.getName());
```
Use `JcrConstants.JCR_TITLE` and rather import `com.day.cq.commons.jcr.JcrConstants` instead of the old jackrabbit version.

cjelger commented on bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/gui/components/configuration/ConfigurationColumnPreview.java:
```diff
@@ -0,0 +1,91 @@
+        LOG.debug("Item in preview is at path {}", itemResourcePath);
+
+        itemResource = request.getResourceResolver().getResource(itemResourcePath);
+        isFolder = itemResource.isResourceType(JcrConstants.NT_FOLDER) || itemResource.isResourceType("sling:Folder") || itemResource
```
If available, you can also use the `sling:` constants from `org.apache.sling.jcr.resource.JcrResourceConstants`

cjelger commented on bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/gui/components/configuration/ConfigurationColumnViewItem.java:
```diff
@@ -0,0 +1,79 @@
+
+    private boolean isConfigurationContainer(Resource res) {
+        return (res.getPath()
+            .startsWith(Constants.CONF_ROOT) && (resource.isResourceType("sling:Folder") || resource.isResourceType("sling:OrderedFolder"))
```
If available, you can also use the `sling:` constants from `org.apache.sling.jcr.resource.JcrResourceConstants`

cjelger commented on bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/gui/components/configuration/ConfigurationColumnViewItem.java:
```diff
@@ -0,0 +1,79 @@
+        return ImmutableList.of("cq-confadmin-actions-properties-activator", "cq-confadmin-actions-delete-activator");
+    }
+
+    private boolean isConfigurationContainer(Resource res) {
```
Why do you pass `resource` as a parameter when it's a private member of the class? You also mix `res` and `resource` ...

cjelger replied:
Would be nice if you could also fix this one here.

cjelger commented on bundles/cif-virtual-catalog/src/main/java/com/adobe/cq/commerce/virtual/catalog/data/impl/CatalogDataResourceProviderManagerImpl.java:
```diff
@@ -68,18 +71,34 @@
     private static final String VIRTUAL_PRODUCTS_SERVICE = "virtual-products-service";
+
     private static final String OBSERVATION_PATHS_DEFAULT = "/var/commerce/products";
+    // private static final String OBSERVATION_PATHS_DEFAULT = "/conf";
```
Please remove the comment, also the one below for the query.

cjelger commented on bundles/cif-virtual-catalog/src/main/java/com/adobe/cq/commerce/virtual/catalog/data/impl/CatalogDataResourceProviderManagerImpl.java:
```diff
@@ -68,18 +71,34 @@
     private String[] findAllQueries = { FINDALLQUERIES_DEFAULT };
-    private String[] obervationPaths = { OBSERVATION_PATHS_DEFAULT };
+
+    private String[] obervationPaths = { OBSERVATION_PATHS_DEFAULT, "/conf" };
```
please rename to `observationPaths`

cjelger commented on bundles/cif-virtual-catalog/src/main/java/com/adobe/cq/commerce/virtual/catalog/data/impl/CatalogDataResourceProviderManagerImpl.java:
```diff
@@ -68,18 +71,34 @@
+
+    private String[] obervationPaths = { OBSERVATION_PATHS_DEFAULT, "/conf" };
+
+    private static final List<String> WATCHED_PROPERTIES = ImmutableList.of("magentoRootCategoryId",
```
If possible, we should have constants for all these hard-coded property names. It's a small list, but I think using a set might be better for the lookups you do.

cjelger commented on bundles/cif-virtual-catalog/src/main/java/com/adobe/cq/commerce/virtual/catalog/data/impl/CatalogDataResourceProviderManagerImpl.java:
```diff
@@ -334,15 +379,16 @@ public void onEvent(EventIterator events) {
                     final Node node = session.getNode(path);
-                    if (node != null && node.isNodeType("sling:Folder") &&
-                        node.hasProperty(CatalogDataResourceProviderFactory.PROPERTY_FACTORY_ID)) {
+                    if (node != null && node.isNodeType("sling:Folder") && node.hasProperty(
```
Use constants, maybe also for `cq:conf` that you actually use in multiple places.

cjelger commented on bundles/cif-virtual-catalog/src/test/java/com/adobe/cq/commerce/virtual/catalog/data/impl/CatalogDataResourceProviderManagerImplTest.java:
```diff
@@ -41,6 +41,10 @@
 import com.day.cq.commons.jcr.JcrUtil;
 import junitx.util.PrivateAccessor;
 
+import static org.mockito.Matchers.any;
```
There wasn't any change to that file ... why the new imports?

cjelger commented on content/cif-connector/src/main/content/jcr_root/apps/commerce/gui/content/products/bindproducttreewizard/customfields/cqConf/.content.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I had a quick check with Mark, actually you can completely remove that `customfields` part: this was made to add specific fields only for the Magento config (storeview and cat id), but we don't need that any more. You could just add that field directly in the wizard panel.

herzog31 replied:
+1

herzog31 commented:
@dplaton Please also add some documentation of all the changes. Like how do new customers configure the service and how do existing customers migrate.

dplaton committed:
Follow-up on the review:

- remove the "custom fields" section from the product binding wizard and put the "cq:conf" field directly in the wizard. Also, update the content package filters.
- format copyright headers
- move the Constants to the `cif-virtual-catalog` bundle since it's used there as well.

dplaton committed:
Implement unit tests for the Sling Models.

The testing dependencies had to be trimmed a bit because of some clashing versions.

dplaton committed:
@trivial Fix formatting

dplaton committed:
Fix some bad code.

dplaton committed:
Update README with the instructions

cjelger commented on bundles/cif-connector-graphql/pom.xml:
```diff
@@ -243,22 +266,37 @@
+            <version>2.3.18</version>
+            <scope>test</scope>
         </dependency>
+<!--        <dependency>-->
```
remove commented dependency if no longer needed

dplaton committed:
Remove unused dependency

cjelger committed:
Merge branch 'master' into issue/CIF-1134

cjelger committed:
CIF-1134 - Unified configuration of site settings

- fix git merge issue

dplaton committed:
Update the "Catalog identifier" field in the configuration properties to be a drop-down instead of a text field.
Add the "Language" field in the configuration dialog.

dplaton committed:
Merge branch 'master' into issue/CIF-1134

dplaton committed:
@trivial Fix formatting

LSantha committed:
Merge branch 'master' into issue/CIF-1134

dplaton committed:
Follow-up on the review:

- remove the confirmation dialog after a configuration is created
- add checks for `null` to remove the "Failed to load data" error which is displayed after deleting a configuration

dplaton committed:
Follow-up on the review:

- replace the "Delete configuration" confirmation dialog so that it doesn't have the "archive page" checkbox

dplaton committed:
Replace the ugly folder icon in column preview

dplaton committed:
Use proper icons in the column preview.

Implement a "client side render condition" to hide the `Create configuration` action if the configuration bucket already has a "commerce" configuration.

dplaton committed:
Don't display the "Create configuration" action if no column is active in the console.

dplaton committed:
Fix the "Cancel" button that didn't do anything.

dplaton committed:
@trivial Fix formatting

dplaton committed:
Fix the issue which was causing the "Create configuration" wizard to not save the data.
Update the precondition for the "Create configuration" action to display only when the configuration has a `settings/cloudconfigs` bucket.

dplaton committed:
Create another unit test to test reading the configuration from `/conf`.

The `CatalogDataProviderManagerImplTest` is perfectly fine, but it uses the old Granite testing library which makes it harder to use the Context Aware feature.

dplaton committed:
Update the unit test to also attempt to register an invalid data root.

dplaton committed:
Update the unit test to also attempt to register an invalid data root.

dplaton committed:
@trivial Fix formatting

LSantha committed:
CIF-1204 - improved code coverage

dplaton committed:
Fix an issue in the properties page which caused the "Catalog identifier" field to be empty.

dplaton committed:
Merge branch 'issue/CIF-1134' of github.com:adobe/commerce-cif-connector into issue/CIF-1134

dplaton committed:
@trivial Fix formatting
